File: proof_kernel/run_presets.py
"""Research run preset contracts.

Presets declare intent and policy. They do not name concrete data files;
the data planner resolves the strongest valid assets at preflight time.
"""

from __future__ import annotations

import copy
import json
from typing import Any, Dict, List


RUN_PRESET_IDS = ("smoke", "standard", "authoritative", "overkill", "overkill_full")


_BASE_POLICIES: Dict[str, Any] = {
    "zero_policy": {
        "selection": "highest_available",
        "allow_lower_precision_fallback": True,
        "require_odlyzko_crosscheck": False,
    },
    "tau_policy": {
        "selection": "highest_available",
        "require_dps_plus_guard": True,
    },
    "prime_policy": {
        "selection": "canonical_7m",
        "require_sufficient_max_prime": True,
    },
    "certificate_policy": {
        "require_raw_high_precision_artifacts": False,
    },
    "runtime_policy": {
        "run": "all",
        "quick": False,
        "resolution": 500,
        "x_start": None,
        "x_end": None,
        "k_values": "0,1,2",
        "n_test": 500,
        "prime_min_count": 0,
        "prime_target_count": 0,
    },
}


def _contract(
    *,
    preset: str,
    requested_dps: int,
    requested_zero_count: int,
    guard_dps: int,
    allow_lower_precision_fallback: bool,
    require_odlyzko_crosscheck: bool,
    require_raw_high_precision_artifacts: bool,
    runtime_policy: Dict[str, Any] | None = None,
) -> Dict[str, Any]:
    out = copy.deepcopy(_BASE_POLICIES)
    out.update({
        "preset": preset,
        "requested_dps": int(requested_dps),
        "requested_zero_count": int(requested_zero_count),
        "guard_dps": int(guard_dps),
    })
    out["zero_policy"]["allow_lower_precision_fallback"] = bool(allow_lower_precision_fallback)
    out["zero_policy"]["require_odlyzko_crosscheck"] = bool(require_odlyzko_crosscheck)
    out["certificate_policy"]["require_raw_high_precision_artifacts"] = bool(
        require_raw_high_precision_artifacts
    )
    if runtime_policy:
        out["runtime_policy"].update(runtime_policy)
    return out


_PRESETS: Dict[str, Dict[str, Any]] = {
    "smoke": _contract(
        preset="smoke",
        requested_dps=30,
        requested_zero_count=100,
        guard_dps=0,
        allow_lower_precision_fallback=True,
        require_odlyzko_crosscheck=False,
        require_raw_high_precision_artifacts=False,
        runtime_policy={"quick": True},
    ),
    "standard": _contract(
        preset="standard",
        requested_dps=40,
        requested_zero_count=2000,
        guard_dps=20,
        allow_lower_precision_fallback=True,
        require_odlyzko_crosscheck=False,
        require_raw_high_precision_artifacts=False,
    ),
    "authoritative": _contract(
        preset="authoritative",
        requested_dps=80,
        requested_zero_count=100000,
        guard_dps=20,
        allow_lower_precision_fallback=False,
        require_odlyzko_crosscheck=True,
        require_raw_high_precision_artifacts=True,
        runtime_policy={"prime_min_count": 1000000, "prime_target_count": 1000000},
    ),
    "overkill": _contract(
        preset="overkill",
        requested_dps=80,
        requested_zero_count=100000,
        guard_dps=20,
        allow_lower_precision_fallback=False,
        require_odlyzko_crosscheck=True,
        require_raw_high_precision_artifacts=True,
        runtime_policy={"prime_min_count": 1000000, "prime_target_count": 1000000},
    ),
    # Back-compat extension retained for existing UI/API callers. The preset
    # contract is the same as overkill except for requesting the full prime file.
    "overkill_full": _contract(
        preset="overkill_full",
        requested_dps=80,
        requested_zero_count=100000,
        guard_dps=20,
        allow_lower_precision_fallback=False,
        require_odlyzko_crosscheck=True,
        require_raw_high_precision_artifacts=True,
        runtime_policy={
            "prime_min_count": 1000000,
            "prime_target_count": 7000000,
            "x_start": 2,
            "x_end": 50,
        },
    ),
}


def is_run_preset(value: Any) -> bool:
    return str(value or "").lower() in _PRESETS


def resolve_run_preset(preset: str = "standard", **overrides: Any) -> Dict[str, Any]:
    preset_id = str(preset or "standard").lower()
    if preset_id not in _PRESETS:
        raise ValueError(f"Unknown run preset: {preset}")
    contract = copy.deepcopy(_PRESETS[preset_id])
    for key in ("requested_dps", "requested_zero_count", "guard_dps"):
        if key in overrides and overrides[key] is not None:
            contract[key] = int(overrides[key])
    for policy_key in ("zero_policy", "tau_policy", "prime_policy", "certificate_policy", "runtime_policy"):
        patch = overrides.get(policy_key)
        if isinstance(patch, dict):
            contract[policy_key].update(copy.deepcopy(patch))
    return contract


def get_run_presets() -> List[Dict[str, Any]]:
    return [resolve_run_preset(preset_id) for preset_id in RUN_PRESET_IDS]


if __name__ == "__main__":
    print(json.dumps({"presets": get_run_presets()}, indent=2, sort_keys=True))

File: lib/run-presets.ts
export type RunPresetId = "smoke" | "standard" | "authoritative" | "overkill" | "overkill_full";

export interface RunPresetContract {
    preset: RunPresetId;
    requested_dps: number;
    requested_zero_count: number;
    guard_dps: number;
    zero_policy: {
        selection: "highest_available";
        allow_lower_precision_fallback: boolean;
        require_odlyzko_crosscheck: boolean;
        crosscheck_tolerance?: string;
    };
    tau_policy: {
        selection: "highest_available";
        require_dps_plus_guard: boolean;
    };
    prime_policy: {
        selection: "canonical_7m";
        require_sufficient_max_prime: boolean;
    };
    certificate_policy: {
        require_raw_high_precision_artifacts: boolean;
    };
    runtime_policy: {
        run: string;
        quick: boolean;
        resolution: number;
        x_start: number | null;
        x_end: number | null;
        k_values: string;
        n_test: number;
        prime_min_count: number;
        prime_target_count: number;
    };
}

const basePolicies = () => ({
    zero_policy: {
        selection: "highest_available" as const,
        allow_lower_precision_fallback: true,
        require_odlyzko_crosscheck: false,
    },
    tau_policy: {
        selection: "highest_available" as const,
        require_dps_plus_guard: true,
    },
    prime_policy: {
        selection: "canonical_7m" as const,
        require_sufficient_max_prime: true,
    },
    certificate_policy: {
        require_raw_high_precision_artifacts: false,
    },
    runtime_policy: {
        run: "all",
        quick: false,
        resolution: 500,
        x_start: null as number | null,
        x_end: null as number | null,
        k_values: "0,1,2",
        n_test: 500,
        prime_min_count: 0,
        prime_target_count: 0,
    },
});

const contract = (
    preset: RunPresetId,
    requestedDps: number,
    requestedZeroCount: number,
    guardDps: number,
    options: {
        allowLowerPrecisionFallback: boolean;
        requireOdlyzkoCrosscheck: boolean;
        requireRawHighPrecisionArtifacts: boolean;
        runtimePolicy?: Partial<RunPresetContract["runtime_policy"]>;
    },
): RunPresetContract => {
    const base = basePolicies();
    return {
        preset,
        requested_dps: requestedDps,
        requested_zero_count: requestedZeroCount,
        guard_dps: guardDps,
        zero_policy: {
            ...base.zero_policy,
            allow_lower_precision_fallback: options.allowLowerPrecisionFallback,
            require_odlyzko_crosscheck: options.requireOdlyzkoCrosscheck,
        },
        tau_policy: base.tau_policy,
        prime_policy: base.prime_policy,
        certificate_policy: {
            require_raw_high_precision_artifacts: options.requireRawHighPrecisionArtifacts,
        },
        runtime_policy: {
            ...base.runtime_policy,
            ...(options.runtimePolicy ?? {}),
        },
    };
};

const PRESETS: Record<RunPresetId, RunPresetContract> = {
    smoke: contract("smoke", 30, 100, 0, {
        allowLowerPrecisionFallback: true,
        requireOdlyzkoCrosscheck: false,
        requireRawHighPrecisionArtifacts: false,
        runtimePolicy: { quick: true },
    }),
    standard: contract("standard", 40, 2000, 20, {
        allowLowerPrecisionFallback: true,
        requireOdlyzkoCrosscheck: false,
        requireRawHighPrecisionArtifacts: false,
    }),
    authoritative: contract("authoritative", 80, 100000, 20, {
        allowLowerPrecisionFallback: false,
        requireOdlyzkoCrosscheck: true,
        requireRawHighPrecisionArtifacts: true,
        runtimePolicy: { prime_min_count: 1_000_000, prime_target_count: 1_000_000 },
    }),
    overkill: contract("overkill", 80, 60000, 20, {
        allowLowerPrecisionFallback: false,
        requireOdlyzkoCrosscheck: true,
        requireRawHighPrecisionArtifacts: true,
        runtimePolicy: { prime_min_count: 1_000_000, prime_target_count: 7_000_000 },
    }),
    overkill_full: contract("overkill_full", 80, 100000, 20, {
        allowLowerPrecisionFallback: false,
        requireOdlyzkoCrosscheck: true,
        requireRawHighPrecisionArtifacts: true,
        runtimePolicy: { prime_min_count: 1_000_000, prime_target_count: 7_000_000, x_start: 2, x_end: 50 },
    }),
};

export const isRunPreset = (value: unknown): value is RunPresetId =>
    typeof value === "string" && value.toLowerCase() in PRESETS;

export const resolveRunPreset = (
    preset: RunPresetId | string = "standard",
    overrides: Partial<RunPresetContract> = {},
): RunPresetContract => {
    const presetId = String(preset).toLowerCase() as RunPresetId;
    const base = PRESETS[presetId];
    if (!base) throw new Error(`Unknown run preset: ${preset}`);
    return {
        ...structuredClone(base),
        ...overrides,
        zero_policy: { ...base.zero_policy, ...(overrides.zero_policy ?? {}) },
        tau_policy: { ...base.tau_policy, ...(overrides.tau_policy ?? {}) },
        prime_policy: { ...base.prime_policy, ...(overrides.prime_policy ?? {}) },
        certificate_policy: { ...base.certificate_policy, ...(overrides.certificate_policy ?? {}) },
        runtime_policy: { ...base.runtime_policy, ...(overrides.runtime_policy ?? {}) },
    };
};

export const getRunPresets = () =>
    (["smoke", "standard", "authoritative", "overkill", "overkill_full"] as RunPresetId[]).map((preset) =>
        resolveRunPreset(preset),
    );
